File: Ruby/bf.rb
data = Array.new(30000, 0)
p = 0
s = ""

def process_brainfuck(loc, data, p, s)
  t = loc
  while t < s.length do
    c = s[t]
    if c == '+'
      data[p] += 1
    elsif c == '-'
      data[p] -= 1
    elsif c == '>'
      p += 1
    elsif c == '<'
      p -= 1
    elsif c == '.'
      print data[p].chr
    elsif c == '['
      t = process_brainfuck(t + 1, data, p, s)
    elsif c == ']' && data[p] != 0
      if data[p] != 0
        return process_brainfuck(loc, data, p, s)
      end
      return t
    end
    t += 1
  end
  return s.length
end

filePath = ARGV[0]
File.open(filePath, 'r') do |file|
  s = file.read
  process_brainfuck(0, data, p, s)
end

File: PHP/bf.php
<?php

$data = array_fill(0, 30000, 0);
$p = 0;
$s = "";

function process_brainfuck($loc) {
  global $data, $p, $s;

  $t = $loc;
  while ($t < strlen($s)) {
    $c = $s[$t];
    if ($c == '+') { $data[$p]++; }
    if ($c == '-') { $data[$p]--; }
    if ($c == '>') { $p++; }
    if ($c == '<') { $p--; }
    if ($c == '.') { echo chr($data[$p]); }
    if ($c == '[') { $t = process_brainfuck($t + 1); }
    if ($c == ']' && $data[$p] != 0) {
      if ($data[$p] != 0) return process_brainfuck($loc);
      return $t;
    }
    $t++;
  }
  return strlen($s);
}

$filePath = $argv[1];
$content = file_get_contents($filePath);
if ($content === false) {
  echo "Error reading file: $filePath";
  return;
}
$s = $content;
process_brainfuck(0);

?>

